### DISCUSSION TITLE
Linker error in Running the MNIST sample app

### DISCUSSION THREAD

**Discussion by mayani-nv** [microsoft/onnxruntime#5209] [Other Q&A] [answered]
I followed the instructions described [here](https://github.com/microsoft/onnxruntime/blob/master/BUILD.md#start-baseline-cpu) to build ORT from source. The build was successful and therefore, I wanted to try the MNIST app given in the samples folder. I used the commands described [here](https://github.com/microsoft/onnxruntime/tree/master/samples/c_cxx/MNIST) to run it but got the following error 

```
C:\Users\Administrator\onnxruntime\samples\c_cxx\MNIST>build.bat

C:\Users\Administrator\onnxruntime\samples\c_cxx\MNIST>cl MNIST.cpp /Zi /EHsc /I..\..\..\include\onnxruntime\core\session /link /LIBPATH:..\..\..\build\Windows\Debug\Debug
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27043 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

MNIST.cpp
Microsoft (R) Incremental Linker Version 14.16.27043.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:MNIST.exe
/debug
/LIBPATH:..\..\..\build\Windows\Debug\Debug
MNIST.obj
MNIST.obj : error LNK2019: unresolved external symbol _OrtGetApiBase@0 referenced in function "void __cdecl `dynamic initializer for 'public: static struct OrtApi const * const Ort::Global<void>::api_''(void)" (??__E?api_@?$Global@X@Ort@@2PBUOrtApi@@B@@YAXXZ)
..\..\..\build\Windows\Debug\Debug\onnxruntime.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
MNIST.exe : fatal error LNK1120: 1 unresolved externals
```
**System information**
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows Server 2016
ONNX Runtime installed from (source or binary): source
Visual Studio version (if applicable): 2017 Community edition
Cmake version: 3.18.2

**Comment by hariharans29** [comment]
It seems like the x86 compiler (cl.exe) is being used to build the MNIST.exe app and that is linking to the onnxruntime.dll built for x64 architecture and that causes the linker warning/error.

![image](https://user-images.githubusercontent.com/9969784/93514774-75bc7500-f8dc-11ea-9e5c-cd0297caef05.png)


Can you try using the `x64 Native Tools Command Prompt`  ?

**Comment by hariharans29** [comment]
https://social.msdn.microsoft.com/Forums/vstudio/en-US/daabb34a-838a-4603-8024-b95cea107594/how-to-set-environment-variables-for-msvc-clexe-compiler-alongside-older-versions?forum=visualstudiogeneral

**Reply by ankur2051** [reply]
Hi,

How would you propose to solve the issue in case of loading such a DLL using python ctypes?
I am working on a project where i have to use the inference of a trained model in c++. I used onnnxruntime and i am to build the DLL but while linking it to the other DLL of the project I am facing issues, as the onnxruntime.dll from the c/windows/system32 is loaded while using ctypes.CDLL() (a project-based requirement). I cannot put the DLL at the location of python.exe as this is a project-level violation. Could you please suggest some alternative?

**Comment by mayani-nv** [comment]
@hariharans29 Thanks for the suggestion. I used the x64 Native tools command prompt and the build command works. However, I get the following error when I type `.\MNIST.exe`. 
![Capture](https://user-images.githubusercontent.com/67936769/93521889-84f3f080-f8e5-11ea-990f-befbcba593ce.PNG)

So I looked that the onnxruntime.dll is located on the path `C:\Users\Administrator\onnxruntime\build\Windows\Debug`. should I add this path in the list of Path environmental variable?

**Comment by hariharans29** [comment]
Yes, your app should be able to find the onnxruntime dll. Pls add it to PATH and try.

**Comment by mayani-nv** [comment]
@hariharans29 Thanks. I am able to run the MNIST.exe but I cannot see the GUI as mentioned in this [tutorial](https://github.com/microsoft/onnxruntime/tree/master/samples/c_cxx/MNIST). Could you please help with that

**Comment by hariharans29** [comment]
I am not familiar with the GUI aspect of the tutorial. 

CC: @snnn who contributed the tutorial and might be able to help

**Comment by snnn** [comment]
Hi @mayani-nv , excuse me, instead of setting PATH, you should copy onnxruntime.dll to the same folder of MNIST.exe.  In general, set PATH is not a reliable way to resolve DLL not found issues.

**Comment by mayani-nv** [comment]
@snnn Thanks for the suggestion. I copied the onnxruntime.dll file to the MNIST folder. It gets build but I got the following message. 
```
C:\Users\Administrator\onnxruntime\samples\c_cxx\MNIST>.\build.bat

C:\Users\Administrator\onnxruntime\samples\c_cxx\MNIST>cl MNIST.cpp /Zi /EHsc /I..\..\..\include\onnxruntime\core\session /link /LIBPATH:..\..\..\build\Windows\Debug\Debug
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27043 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

MNIST.cpp
Microsoft (R) Incremental Linker Version 14.16.27043.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:MNIST.exe
/debug
/LIBPATH:..\..\..\build\Windows\Debug\Debug
MNIST.obj
LINK : MNIST.exe not found or not built by the last incremental link; performing full link

C:\Users\Administrator\onnxruntime\samples\c_cxx\MNIST>.\MNIST.exe

C:\Users\Administrator\onnxruntime\samples\c_cxx\MNIST> ``
````
I still cannot see the GUI

**Comment by snnn** [comment]
Debugging.

**Comment by snnn** [comment, marked as answer]
It works fine on my PC.  Based on the limited information I have now, it's hard to say what went wrong in your place. 

I have two suggestions:
1.  Download the latest onnxruntime binary release from  https://github.com/microsoft/onnxruntime/releases/tag/v1.4.0, the "onnxruntime-win-x64-1.4.0.zip" file. Then unzip it, and modify the paths in build.bat you have as something like:
```
cl MNIST.cpp /Zi /EHsc /IC:\src\onnxruntime\samples\c_cxx\MNIST\onnxruntime-win-x64-1.4.0\include /link /LIBPATH:C:\src\onnxruntime\samples\c_cxx\MNIST\onnxruntime-win-x64-1.4.0\lib
```
Then run it and copy the onnxruntime.dll from unzipped folder to the same dir of MNIST.exe. Also, please don't forget to download the model file and put it there. 

2. Use the instructions at https://github.com/microsoft/onnxruntime/tree/master/samples/c_cxx instead.  You'll get a Visual C++ solution file and you can open in the Visual Studio graphical interface, which have a nice debugger. Then you can setup breakpoints on C/C++ exceptions and tell me what where it stopped.

**Comment by mayani-nv** [comment]
@snnn Thanks for the suggestion. The option 1 works and now I can see the GUI. Very much appreciate your help.
